bittide/src/Bittide/Transceiver/WordAlign.hs

```diff
@@ -0,0 +1,178 @@
+splitMsbs ::
+  forall nBytes byteWidth .
+  ( KnownNat nBytes
+  , KnownNat byteWidth
+  , 1 <= byteWidth
+  ) =>
+  BitVector (nBytes * byteWidth) ->
+  ( BitVector nBytes
+  , BitVector (nBytes * (byteWidth - 1))
+  )
+splitMsbs =
+    bimap pack pack
+  . unzip
+  . unpack @(Vec nBytes (Bit, BitVector (byteWidth - 1)))
+
+-- | Opposite of 'splitMsbs'.
+joinMsbs ::
+  forall nBytes byteWidth .
+  ( KnownNat nBytes
+  , KnownNat byteWidth
+  , 1 <= byteWidth
+  ) =>
+  BitVector nBytes ->
+  BitVector (nBytes * (byteWidth - 1)) ->
+  BitVector (nBytes * byteWidth)
+joinMsbs msbs bvs = pack $
+  zip
+    (unpack @(Vec nBytes Bool) msbs)
+    (unpack @(Vec nBytes (BitVector(byteWidth - 1))) bvs)
+
+-- | Specialized version of 'align' that works on 'BitVector' and assumes that
+-- the alignment bits are stored in the MSBs of the bytes.
+alignBytesFromMsbs ::
+  forall n dom .
+  ( HiddenClock dom
+  , KnownNat n
+  , 1 <= n
+  ) =>
+  -- | Freeze alignment
+  Signal dom Bool ->
+  -- | Data with alignment bits in the byte's MSBs
+  Signal dom (Bytes n) ->
+  -- | 'Nothing' until an alignment symbol has been observed. 'Just' if an alignment
+  -- symbol has been observed.
+  Signal dom (Bytes n)
+alignBytesFromMsbs freeze dat =
+  let msbs = fst . splitMsbs @n <$> dat
+   in align freeze (unpack <$> msbs) dat
+
+-- | Align data according to an alignment signal. The alignment signal is a
+-- one-hot encoded signal that indicates which bit is the alignment bit. Note
+-- that the behavior of this function is undefined when:
+--
+--   * Switching from one alignment bit to another alignment bit
+--   * No alignment bit or multiple are asserted
+--
+-- Alignment can be "frozen" by asserting the freeze signal. This is useful when
+-- switching over from a handshake protocol to a data protocol.
+align ::
+  forall n m dom .
+  ( HiddenClock dom
+  , KnownNat n
+  , KnownNat m
+  , 1 <= n
+  ) =>
+  -- | Freeze alignment
+  Signal dom Bool ->
+  -- | One hot encoded alignment signal. Behavior is undefined when multiple
+  -- bits or none are asserted.
+  Signal dom (Vec n Bool) ->
+  -- | Data
+  Signal dom (BitVector (n * m)) ->
+  -- | 'Nothing' until an alignment symbol has been observed. 'Just' if an
+  -- alignment symbol has been observed.
+  Signal dom (BitVector (n * m))
+align =
+  withEnable enableGen $
+    withReset noReset $
+      curry3 (mealy go (defState @n @m) . bundle)
+ where
+  go :: State n m -> (Bool, Vec n Bool, BitVector (n * m)) -> (State n m, BitVector (n * m))
+  go (State{prev, offset}) (freeze, alignment, current) =
+    ( State{prev=current, offset=newOffset}
+    , takeMsbs (shiftL (prev ++# current) (fromIntegral (offset `mul` mAsIndex)))
+    )
+   where
+    mAsIndex = maxBound :: Index (m + 1)
+
+    newOffset
+      | freeze = offset
+      | otherwise = fromMaybe 0 (elemIndex True alignment)
+
+  takeMsbs :: forall i . KnownNat i => BitVector (i * 2) -> BitVector i
+  takeMsbs bv = truncateB (rotateR bv (natToNum @i))
```

Comment:
Is there a particular reason that we rotate here? Wouldn't `shiftR` be sufficient already?

Reply:
Yep, doesn't make sense for sure!